[Cocoa] Copying and pasting adaptive image glyphs across Mail compose windows drops content

pxlcoder opened the pull request:
#### 9456228689d44ef687ffd5782a5323f07c6f233d
<pre>
[Cocoa] Copying and pasting adaptive image glyphs across Mail compose windows drops content
<a href="https://bugs.webkit.org/show_bug.cgi?id=278256">https://bugs.webkit.org/show_bug.cgi?id=278256</a>
<a href="https://rdar.apple.com/130606666">rdar://130606666</a>

Reviewed by Sihui Liu and Wenson Hsieh.

When copying and pasting web content across Mail compose windows, content is
serialized into a web archive. The archive only contains subresources which
have already been loaded.

The above is problematic for adaptive image glyphs, which are represented as a
&lt;picture&gt; element containing a HEIC and a fallback PNG. Since the HEIC is selected
during &lt;source&gt; matching, the fallback PNG never gets loaded. Consequently,
when saving content as a web archive, the fallback PNG is dropped.

Web archives should contain both adaptive image glyphs resources. However, the
fallback data is unavailable in the web content process, and a sync IPC to the
network process is undesirable when saving to a web archive. As a result, fix
by always loading both the HEIC and fallback PNG for adaptive image glyphs, unlike
regular source selection. This ensures both resources are always available.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):

Always load the fallback PNG for adaptive image glyphs, in addition to the
selected source (HEIC).

* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsCopyFromWebViewPasteToWebView)):

Canonical link: <a href="https://commits.webkit.org/282454@main">https://commits.webkit.org/282454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce9f48fa14625b24de46f9ca9d41e46dd4c59eda

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15771 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67195 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13782 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14062 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50920 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9526 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66243 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39497 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31600 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12050 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12654 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57723 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68891 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7121 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11999 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58230 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54772 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58445 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14002 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5942 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38351 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39430 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40542 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -267,6 +267,23 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+                auto request = createPotentialAccessControlRequest(WTFMove(resourceRequest), WTFMove(options), document, crossOriginAttribute);
+                request.setInitiator(*imageElement);
+
+                document->protectedCachedResourceLoader()->requestImage(WTFMove(request));
```
is the return value of `requestImage` not important?

whsieh replied:
I don't think so — the commit message explains that we only need to

> [ensure] both resources are always available.

pxlcoder replied:
yeah, we're not actually drawing this resource

pxlcoder committed:
[Cocoa] Copying and pasting adaptive image glyphs across Mail compose windows drops content
https://bugs.webkit.org/show_bug.cgi?id=278256
rdar://130606666

Reviewed by Sihui Liu and Wenson Hsieh.

When copying and pasting web content across Mail compose windows, content is
serialized into a web archive. The archive only contains subresources which
have already been loaded.

The above is problematic for adaptive image glyphs, which are represented as a
<picture> element containing a HEIC and a fallback PNG. Since the HEIC is selected
during <source> matching, the fallback PNG never gets loaded. Consequently,
when saving content as a web archive, the fallback PNG is dropped.

Web archives should contain both adaptive image glyphs resources. However, the
fallback data is unavailable in the web content process, and a sync IPC to the
network process is undesirable when saving to a web archive. As a result, fix
by always loading both the HEIC and fallback PNG for adaptive image glyphs, unlike
regular source selection. This ensures both resources are always available.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):

Always load the fallback PNG for adaptive image glyphs, in addition to the
selected source (HEIC).

* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsCopyFromWebViewPasteToWebView)):

Canonical link: https://commits.webkit.org/282454@main
